File: cpp/297.cpp
#include <string>
#include <iostream>
#include <sstream>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec {
  TreeNode* d(istringstream &istr){
    if(istr.peek() == -1 || istr.peek() == ',' || istr.peek() == ')'){
      return nullptr;
    } else {
      int v; istr.get(); istr >> v; istr.get();
      TreeNode *t = new TreeNode(v);
      t->left = d(istr); istr.get();
      t->right = d(istr); istr.get();
      return t;
    }
  }

public:

  // Encodes a tree to a single string.
  string serialize(TreeNode* root) {
    if(root == nullptr)
      return "";
    ostringstream ostr;
    ostr << "(" << root->val << ","
         << serialize(root->left) << ","
         << serialize(root->right) << ")";
    return ostr.str();
  }

  // Decodes your encoded data to tree.
  TreeNode* deserialize(string data) {
    istringstream istr(data);
    return d(istr);
  }
};

int main(){
  Codec codec;
  TreeNode *root = new TreeNode(-1);
  root->left = new TreeNode(0);
  root->right = new TreeNode(1);
  cout << codec.serialize(root) << endl;
  codec.deserialize(codec.serialize(root));
}

File: python/297.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        if root is None:
            return "None"
        return "({},{},{})".format(
            root.val,
            self.serialize(root.left),
            self.serialize(root.right))

    @staticmethod
    def _d(s, i):
        if s[i] == 'N':
            return None, i+4
        else:
            j = s.index(',', i)
            t = TreeNode(int(s[i+1:j]))
            t.left, j = Codec._d(s, j+1)
            t.right, j = Codec._d(s, j+1)
            return t, j+1

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        return self._d(data, 0)[0]
